build-tools/Xamarin.Android.Tools.BootstrapTasks/Xamarin.Android.Tools.BootstrapTasks/CreateFrameworkList.cs

```diff
@@ -0,0 +1,84 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Xml.Linq;
+
+using System.Reflection.Metadata;
+using System.Reflection.PortableExecutable;
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace Xamarin.Android.Tools.BootstrapTasks
+{
+	public class CreateFrameworkList : Task
+	{
+		[Required]
+		public                  ITaskItem       FrameworkListFile           { get; set; }
+
+		[Required]
+		public                  ITaskItem       FrameworkDirectory          { get; set; }
+
+		[Required]
+		public                  ITaskItem[]     FrameworkFiles              { get; set; }
+
+		public                  ITaskItem[]     FrameworkFileOverrides      { get; set; }
+
+		[Required]
+		public                  string          Redist                      { get; set; }
+
+		[Required]
+		public                  string          Name                        { get; set; }
+
+		public override bool Execute()
+		{
+			var files = new List<ITaskItem>(FrameworkFiles);
+			files.Sort ((x, y) => {
+					var a = Path.GetFileNameWithoutExtension (x.ItemSpec);
+					var b = Path.GetFileNameWithoutExtension (y.ItemSpec);
+					return string.Compare (a, b, StringComparison.OrdinalIgnoreCase);
+			});
+			var contents = new XElement ("FileList",
+					new XAttribute ("Redist", Redist),
+					new XAttribute ("Name", Name),
+					files.Select (f => ToFileElement (f)));
+			contents.Save (FrameworkListFile.ItemSpec);
+			Console.WriteLine ("# JONP TERRIBLE DEBUGGING MESSAGE!");
+			Console.WriteLine (contents.ToString());
+			return true;
+		}
+
+		AssemblyName    AssemblyName    = new AssemblyName();
+
+		XElement ToFileElement (ITaskItem file)
+		{
+			var path = Path.Combine (FrameworkDirectory.ItemSpec, file.ItemSpec);
+			if (!File.Exists (path)) {
+				path = Path.Combine (FrameworkDirectory.ItemSpec, "Facades", file.ItemSpec);
+			}
+			using (var pe = new PEReader (File.OpenRead (path))) {
+				var reader          = pe.GetMetadataReader();
+				var assemblyDef     = reader.GetAssemblyDefinition();
+
+				var taskVersion     = Nullable (file.GetMetadata ("Version"));
+				var overrideVersion = Nullable (FrameworkFileOverrides.FirstOrDefault (o => o.ItemSpec == file.ItemSpec)?.GetMetadata ("Version"));
+				var assemblyVersion = assemblyDef.Version.ToString ().TrimStart ('v');
+				var version         = taskVersion ?? overrideVersion ?? assemblyVersion;
+
+				AssemblyName.SetPublicKey (reader.GetBlobBytes (assemblyDef.PublicKey));
+
+				var publicKeyToken  = string.Join ("", AssemblyName.GetPublicKeyToken ().Select(b => b.ToString ("x2")));
+
+				return new XElement ("File",
+						new XAttribute ("AssemblyName", Path.GetFileNameWithoutExtension (file.ItemSpec)),
+						new XAttribute ("Version", version),
+						new XAttribute ("PublicKeyToken", publicKeyToken),
+						new XAttribute ("ProcessorArchitecture", "MSIL"));
+			}
```

Comment:
In VS Mac we just use AssemblyName.GetAssemblyName (path/to/assembly) which gives us the public key, processor architecture and assembly version. Which seems simpler than using the PEReader but maybe I am missing something.

Reply:
My normal concern with `System.Reflection` is that it doesn't work with `mscorlib.dll`, so I immediately discounted anything in `System.Reflection`.

Fortunately `AssemblyName.GetAssemblyName()` *does* work with mscorlib.dll:

```csharp
var an = AssemblyName.GetAssemblyName("bin/Debug//lib/xamarin.android/xbuild-frameworks/MonoAndroid/v1.0/mscorlib.dll");
an.ToString();
// mscorlib, Version=2.0.5.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e
```

I'll do that.